Formula/coredns.rb

```diff
@@ -0,0 +1,83 @@
+class Coredns < Formula
+  desc "DNS server that chains plugins"
+  homepage "https://coredns.io/"
+  url "https://github.com/coredns/coredns/archive/v1.6.9.tar.gz"
+  sha256 "e100a946d5d60d936cb313bdb04fd96162d5a7229e08887ad497660acee4b36f"
+  head "https://github.com/coredns/coredns.git"
+
+  depends_on "go" => :build
+
+  def default_config
+    <<~EOS
+      . {
+        hosts {
+          fallthrough
+        }
+        forward . 1.1.1.1 1.0.0.1 8.8.8.8 8.8.4.4 9.9.9.9 149.112.112.112
+        cache
+        errors
+        log
+      }
+    EOS
+  end
+
+  def install
+    system "make"
+    (buildpath/"Corefile.example").write default_config
+    (etc/"coredns").mkpath
+    etc.install "Corefile.example" => "coredns/Corefile"
+    bin.install "coredns"
+  end
+
+  def caveats
+    <<~EOS
+      Config file: #{etc}/coredns/Corefile
+      DNS log file: #{var}/log/coredns.log
+      Start service: sudo brew services start coredns
+      Activate DNS: networksetup -setdnsservers Wi-Fi 127.0.0.1
+      Test functionality: dig google.com @127.0.0.1
+
+      DNS proxy servers:
+
+      Cloudflare -> https://1.1.1.1/
+      Google DNS -> https://dns.google.com/
+      Google DNS -> https://developers.google.com/speed/public-dns
+      IBM Threat -> https://www.quad9.net/faq/
+    EOS
+  end
+
+  plist_options :startup => true
+
+  def plist
+    <<~EOS
+      <?xml version="1.0" encoding="UTF-8"?>
+      <!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+      <plist version="1.0">
+        <dict>
+          <key>Label</key>
+          <string>#{plist_name}</string>
+          <key>ProgramArguments</key>
+          <array>
+            <string>#{opt_bin}/coredns</string>
+            <string>-conf</string>
+            <string>#{etc}/coredns/Corefile</string>
+          </array>
+          <key>RunAtLoad</key>
+          <true/>
+          <key>KeepAlive</key>
+          <true/>
+          <key>StandardErrorPath</key>
+          <string>#{var}/log/coredns.log</string>
+          <key>StandardOutPath</key>
+          <string>#{var}/log/coredns.log</string>
+          <key>WorkingDirectory</key>
+          <string>#{HOMEBREW_PREFIX}</string>
+        </dict>
+      </plist>
+    EOS
+  end
+
+  test do
+    assert_match "CoreDNS-#{version}", shell_output("#{bin}/coredns -version")
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
I have added a realistic test with [commit f147636](https://github.com/Homebrew/homebrew-core/pull/53755/commits/f1476367b58b24c301f5172805c64b7b457968cd) that starts a single instance of CoreDNS on port 6053, waits a few seconds and then sends a DNS query to check if the response matches a specific pattern, then the server is gracefully terminated using SIGINT.